subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/modulecache/DefaultCachedMetadata.java

```diff
@@ -73,7 +79,13 @@ public void putProcessedMetadata(int hash, ModuleComponentResolveMetadata proces
             processedMetadataByRules = Collections.singletonMap(hash, processed);
             return;
         } else if (processedMetadataByRules.size() == 1) {
-            processedMetadataByRules = Maps.newHashMap(processedMetadataByRules);
+            synchronized (this) {
+                if (processedMetadataByRules.size() == 1) {
+                    processedMetadataByRules = new ConcurrentHashMap<>(processedMetadataByRules);
+                    processedMetadataByRules.put(hash, processed);
+                    return;
+                }
+            }
```

Comment:
I would avoid double-checked locking and `volatile` fields unless it's really necessary. It can so easily lead to subtle concurrency bugs. Do we know there is a reason not to simply synchronize on `putProcessedMetadata()`?

Reply:
I'll push an updated version that does just that, as it will protect against another race though  not a correctness issue.
And if perf tests do not flag this change as a regression, it will indeed be cleaner.
Note that because the map is not a final field, it has to remain `volatile` as I really do not want to synchronize the get part.